File: backend/app/schemas/group.py
"""
组相关的 Pydantic 模式
"""
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


class GroupMemberInfo(BaseModel):
    """组成员信息"""
    id: int
    username: str
    real_name: str
    email: Optional[str] = None
    phone: Optional[str] = None
    
    class Config:
        from_attributes = True


class GroupBase(BaseModel):
    """组基础模式"""
    name: str
    description: Optional[str] = None


class GroupCreate(GroupBase):
    """创建组模式"""
    leader_ids: Optional[List[int]] = []  # 组长ID列表（支持多个组长）
    member_ids: Optional[List[int]] = []


class GroupUpdate(BaseModel):
    """更新组模式"""
    name: Optional[str] = None
    description: Optional[str] = None
    leader_ids: Optional[List[int]] = None  # 组长ID列表（支持多个组长）


class GroupResponse(GroupBase):
    """组响应模式"""
    id: int
    leader_id: Optional[int] = None  # 主组长ID（向后兼容）
    leader_name: Optional[str] = None  # 主组长名称（向后兼容）
    leader_ids: Optional[List[int]] = []  # 所有组长ID列表
    leader_names: Optional[List[str]] = []  # 所有组长名称列表
    member_count: int = 0
    members: List[GroupMemberInfo] = []
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True
    
    @classmethod
    def from_orm_with_relations(cls, group):
        """从ORM对象创建响应对象，包含关联数据"""
        from app.models.group import group_members
        from app.models.user import User
        
        # 获取主组长信息（向后兼容）
        leader_name = None
        if group.leader:
            leader_name = group.leader.real_name or group.leader.username
        
        # 获取所有组长信息
        leader_ids = []
        leader_names = []
        if hasattr(group, 'leaders') and group.leaders:
            for leader in group.leaders:
                leader_ids.append(leader.id)
                leader_names.append(leader.real_name or leader.username)
        
        # 如果没有组长列表但有主组长，将主组长添加到列表（向后兼容）
        if not leader_ids and group.leader_id:
            leader_ids.append(group.leader_id)
            if leader_name:
                leader_names.append(leader_name)
        
        # 获取成员列表
        members = []
        if hasattr(group, 'members'):
            for member in group.members:
                members.append(GroupMemberInfo(
                    id=member.id,
                    username=member.username,
                    real_name=member.real_name,
                    email=member.email,
                    phone=member.phone
                ))
        
        return cls(
            id=group.id,
            name=group.name,
            description=group.description,
            leader_id=group.leader_id,  # 保留主组长ID（向后兼容）
            leader_name=leader_name,  # 保留主组长名称（向后兼容）
            leader_ids=leader_ids,  # 所有组长ID列表
            leader_names=leader_names,  # 所有组长名称列表
            member_count=len(members),
            members=members,
            created_at=group.created_at,
            updated_at=group.updated_at
        )


class GroupMemberAdd(BaseModel):
    """添加组成员模式"""
    user_ids: List[int]


class GroupMemberRemove(BaseModel):
    """移除组成员模式"""
    user_ids: List[int]

File: frontend/src/services/group.ts
/**
 * 组管理服务
 */
import apiClient from './api'
import { API_ENDPOINTS } from '../config/api'
import { PaginatedResponse } from '../types/common'

export interface GroupMemberInfo {
  id: number
  username: string
  real_name: string
  email?: string
  phone?: string
}

export interface Group {
  id: number
  name: string
  description?: string
  leader_id?: number  // 主组长ID（向后兼容）
  leader_name?: string  // 主组长名称（向后兼容）
  leader_ids?: number[]  // 所有组长ID列表
  leader_names?: string[]  // 所有组长名称列表
  member_count: number
  members: GroupMemberInfo[]
  created_at: string
  updated_at: string
}

export interface GroupCreate {
  name: string
  description?: string
  leader_ids?: number[]  // 组长ID列表（支持多个组长）
  member_ids?: number[]
}

export interface GroupUpdate {
  name?: string
  description?: string
  leader_ids?: number[]  // 组长ID列表（支持多个组长）
}

export interface GroupMemberAdd {
  user_ids: number[]
}

export interface GroupMemberRemove {
  user_ids: number[]
}

export interface GroupListParams {
  search?: string
  my_group?: boolean
  page?: number
  page_size?: number
}

export const groupService = {
  /**
   * 获取组列表（支持分页和搜索）
   */
  getGroups: async (params?: GroupListParams): Promise<PaginatedResponse<Group>> => {
    const queryParams: any = {}
    if (params?.search) queryParams.search = params.search
    if (params?.my_group !== undefined) queryParams.my_group = params.my_group
    if (params?.page) queryParams.page = params.page
    if (params?.page_size) queryParams.page_size = params.page_size
    return apiClient.get(API_ENDPOINTS.GROUPS, { params: queryParams })
  },

  /**
   * 获取组详情
   */
  getGroup: async (id: number): Promise<Group> => {
    return apiClient.get(API_ENDPOINTS.GROUP_DETAIL(id))
  },

  /**
   * 创建组
   */
  createGroup: async (data: GroupCreate): Promise<Group> => {
    return apiClient.post(API_ENDPOINTS.GROUPS, data)
  },

  /**
   * 更新组
   */
  updateGroup: async (id: number, data: GroupUpdate): Promise<Group> => {
    return apiClient.put(API_ENDPOINTS.GROUP_DETAIL(id), data)
  },

  /**
   * 删除组
   */
  deleteGroup: async (id: number): Promise<void> => {
    return apiClient.delete(API_ENDPOINTS.GROUP_DETAIL(id))
  },

  /**
   * 添加组成员
   */
  addGroupMembers: async (id: number, data: GroupMemberAdd): Promise<Group> => {
    return apiClient.post(API_ENDPOINTS.GROUP_ADD_MEMBERS(id), data)
  },

  /**
   * 移除组成员
   */
  removeGroupMembers: async (id: number, data: GroupMemberRemove): Promise<Group> => {
    return apiClient.delete(API_ENDPOINTS.GROUP_REMOVE_MEMBERS(id), { data })
  },

  /**
   * 获取组成员列表
   */
  getGroupMembers: async (id: number): Promise<GroupMemberInfo[]> => {
    return apiClient.get(API_ENDPOINTS.GROUP_MEMBERS(id))
  },

  /**
   * 获取当前用户所属的组信息
   * 如果用户尚未被分配到任何组，返回 null（而不是抛出错误）
   */
  getMyGroup: async (): Promise<Group | null> => {
    return apiClient.get(API_ENDPOINTS.MY_GROUP)
  },
}

